nomad/node_endpoint.go

```diff
@@ -1150,6 +1152,26 @@ func (n *Node) UpdateAlloc(args *structs.AllocUpdateRequest, reply *structs.Gene
 		return fmt.Errorf("must update at least one allocation")
 	}
 
+	// Ensure the node is allowed to update allocs.
+	// The node needs to successfully heartbeat before updating its allocs.
+	nodeID := args.NodeID
+	if nodeID == "" {
+		// COMPAT 1.14
+		// Maintain backwards compatibility with clients that don't set the
+		// NodeID field in the request.
+		nodeID = args.Alloc[0].NodeID
+	}
+	node, err := n.srv.State().NodeByID(nil, nodeID)
```

Comment:
I wasn't sure how to handle this, but I think it's fine to expect that all allocations being sent here are from the same client?

An alternative would be to just skip this check of there is no node ID.

Reply:
Accessing Node ID from allocs will be the default:
https://github.com/hashicorp/nomad/pull/15068#discussion_r1013503431